daliuge-engine/dlg/io.py

```diff
@@ -242,19 +345,41 @@ class MemoryIO(DataIO):
     A DataIO class that reads/write from/into the BytesIO object given at
     construction time
     """
-
-    _desc: io.BytesIO  # TODO: This might actually be a problem
+    # TODO: This might actually be a problem
+    _desc: io.BufferedReader
+    _buf: io.BytesIO
 
     def __init__(self, buf: io.BytesIO, **kwargs):
         super().__init__()
         self._buf = buf
 
+    class BytesIOReader(io.BufferedReader):
+        """
+        A BinaryIO Reader that wraps a BytesIO object for concurrent
+        reading and writing. Closing this reader will not close the BytesIO
+        object and avoid preventing readers observing the same BytesIO object.
+        """
+        _closed = False
+        def __init__(self, raw: io.BytesIO, buffer_size: int = 2048):
+            assert isinstance(raw, io.BytesIO)
+            # NOTE: BytesIO extends BufferedIOBase instead of RawIOBase. Read
+            # and peek operations may return more bytes than requested.
+            super().__init__(raw, buffer_size)  # type: ignore
+
+        @property
+        def closed(self) -> bool:
+            return self.closed
+
+        def close(self) -> None:
+            self._closed = True
+
     def _open(self, **kwargs):
         if self._mode == OpenMode.OPEN_WRITE:
             return self._buf
         elif self._mode == OpenMode.OPEN_READ:
-            # TODO: potentially wasteful copy
-            return io.BytesIO(self._buf.getbuffer())
+            br = MemoryIO.BytesIOReader(self._buf)
+            br.seek(0)
+            return br
```

Comment:
This change is what's currently breaking dynlib tests. From what I understand the dynlib copy app just uses the read method, but for some reason I can't seem to get the test to pass with anything other than a BytesIO object